File: remove_nth_node_from_end.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *slow, *fast;
    int count = 1;  // count from 1
    slow = fast = head;
    while (fast->next != NULL) {
        slow = slow->next;
        fast = fast->next;
        count += 1;
        if (fast->next == NULL) {
            break;
        }
        fast = fast->next;
        count += 1;
        
    }
    n = count - n + 1;
    
    if (n == 1) {
        return head->next;
    }
    count = 2;  // assume that len >= 2
    struct ListNode *pre, *cur;
    pre = head;
    cur = head->next;
    while (count < n) {
        pre = pre->next;
        cur = cur->next;
        count++;
    }
    pre->next = cur->next;
    return head;
}

/*
[1,2,3,4,5]
5
[1,2,3,4,5]
2
[1,2]
2
[1,2]
1
*/

File: remove_nth_node_from_end.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        slow, fast = head, head
        length = 1
        if fast.next is None:
            return None
        
        while fast.next is not None:
            slow = slow.next
            fast = fast.next
            length += 1
            if fast.next is None:
                break
            fast = fast.next
            length += 1
        n = length - n + 1  # the Nth from end -> the Nth from head
        cur, pre = head.next, head
        count = 1
        if n == 1:
            return head.next
        else:
            count += 1
            
        # if length > 5 and n > length / 2:
            # cur, pre = slow.next, slow
            # if odd:
                # count += (length / 2 - 1)
            # else:
                # count += (length / 2 )
        while count != n:
            cur = cur.next
            pre = pre.next
            count += 1
            
        pre.next = cur.next
        return head